src/clock_control/watchdog.rs

```diff
@@ -0,0 +1,227 @@
+//! RTC Watchdog implementation
+//!
+//! # TODO:
+//! - Add convenience methods for configuration
+//! - Consider add default configuration for start with time only
+
+use crate::prelude::*;
+use embedded_hal::watchdog::{Watchdog, WatchdogDisable, WatchdogEnable};
+use esp32::generic::Variant::Val;
+use esp32::rtccntl::wdtconfig0::*;
+use esp32::RTCCNTL;
+
+pub type WatchdogAction = WDT_STG0_A;
+pub type WatchDogResetDuration = WDT_CPU_RESET_LENGTH_A;
+
+const WATCHDOG_UNBLOCK_KEY: u32 = 0x50D83AA1;
+
+const WATCHDOG_BLOCK_VALUE: u32 = 0x89ABCDEF;
+
+pub struct WatchDog {
+    clock_control_config: super::ClockControlConfig,
+}
+
+/// Watchdog configuration
+#[derive(Debug)]
+pub struct WatchdogConfig<
+    T1: Into<MicroSeconds>,
+    T2: Into<MicroSeconds>,
+    T3: Into<MicroSeconds>,
+    T4: Into<MicroSeconds>,
+> {
+    pub period1: T1,
```

Comment:
```suggestion
    pub period1: MicroSeconds,
```

Users of the API should manage conversion for the types, makes it a little bit manageable. Applies to all `periodN` fields.

Reply:
Matter of preference I guess: this way it is very user friendly for the users of the API.

Why wouldn't we do it like this?

Reply:
Because if the user wants to use `ẀatchdogConfig` as an parameter it would need to write every time type, or add type parameters to a function that specify the unit of time for each period in `WatchdogConfig`, or using it in a container. We don't know the use cases for it yet, but.

`period` values are always in milliseconds, and the `MilliSeconds` type should have conversions seconds or from other units of time to `MilliSeconds`.

Reply:
The drawback I see is that when manually construction the WatchdogConfig, you will have to write .into() on all periods:
```
let watchdog_config = esp32_hal::clock_control::watchdog::WatchdogConfig {
        period1: 10.ms().into(),
        action1: esp32_hal::clock_control::watchdog::WatchdogAction::RESETRTC,
        period2: 10.ms().into(),
        action2: esp32_hal::clock_control::watchdog::WatchdogAction::RESETRTC,
        period3: 10.ms().into(),
        action3: esp32_hal::clock_control::watchdog::WatchdogAction::RESETRTC,
        period4: 10.ms().into(),
        action4: esp32_hal::clock_control::watchdog::WatchdogAction::RESETRTC,
        cpu_reset_duration: esp32_hal::clock_control::watchdog::WatchDogResetDuration::T1600NS,
        sys_reset_duration: esp32_hal::clock_control::watchdog::WatchDogResetDuration::T1600NS,
        pause_in_sleep: true,
        reset_cpu: [true, true],
    };
```

Automatic conversion from higher to lower units are present in the units.rs file (e.g. seconds to milliseconds, milliseconds to microseconds.). Not the other way around as these would be lossy.

For the functions and structs I have implemented I chose the lowest unit which can represent the period with full accuracy (in this case microseconds, but for clock ticks it is nanoseconds.

Reply:
>Automatic conversion from higher to lower units are present in the units.rs file (e.g. seconds to milliseconds, milliseconds to microseconds.). Not the other way around as these would be lossy.

It makes sense, that's the reason `WatchdogConfig` will never be returned with `WatchdogConfig<Seconds, ...>`, which in that case doesn't make sense, because precision is loss, and also, you'll end up using anyway `MicroSeconds` under the hood.

I would agree with you to leave this as is if it were only an argument to be passed, which totally makes sense to give the users more ergonomics, but it's also a return type, which the user can store or save, or whichever use they give to it.

If you want to keep this as is there's no problem from me as I don't have an strong opinion more than keeping the flexibility.

Reply:
You make a good case. I will modify.
So going forward we will use generics for function parameters, but not for struct members.